zcash_primitives/src/transaction/components.rs

```diff
@@ -0,0 +1,302 @@
+
+        let mut zkproof = [0; GROTH_PROOF_SIZE];
+        reader.read_exact(&mut zkproof)?;
+        let spend_auth_sig = Signature::read(&mut reader)?;
+
+        Ok(SpendDescription {
+            cv,
+            anchor,
+            nullifier,
+            rk,
+            zkproof,
+            spend_auth_sig,
+        })
+    }
+
+    pub fn write<W: Write>(&self, mut writer: W) -> io::Result<()> {
+        self.cv.write(&mut writer)?;
+        self.anchor.into_repr().write_le(&mut writer)?;
+        writer.write_all(&self.nullifier)?;
+        self.rk.write(&mut writer)?;
+        writer.write_all(&self.zkproof)?;
+        self.spend_auth_sig.write(&mut writer)
+    }
+}
+
+pub struct OutputDescription {
+    pub cv: edwards::Point<Bls12, Unknown>,
+    pub cmu: Fr,
+    pub ephemeral_key: edwards::Point<Bls12, Unknown>,
+    pub enc_ciphertext: [u8; 580],
+    pub out_ciphertext: [u8; 80],
+    pub zkproof: [u8; GROTH_PROOF_SIZE],
+}
+
+impl OutputDescription {
+    pub fn read<R: Read>(mut reader: &mut R) -> io::Result<Self> {
+        let cv = edwards::Point::<Bls12, Unknown>::read(&mut reader, &JUBJUB)?;
+        let cmu = {
+            let mut f = FrRepr::default();
+            f.read_le(&mut reader)?;
+            Fr::from_repr(f).map_err(|e| io::Error::new(io::ErrorKind::InvalidInput, e))?
+        };
+        let ephemeral_key = edwards::Point::<Bls12, Unknown>::read(&mut reader, &JUBJUB)?;
+
+        let mut enc_ciphertext = [0; 580];
+        let mut out_ciphertext = [0; 80];
+        reader.read_exact(&mut enc_ciphertext)?;
+        reader.read_exact(&mut out_ciphertext)?;
+
+        let mut zkproof = [0; GROTH_PROOF_SIZE];
+        reader.read_exact(&mut zkproof)?;
+
+        Ok(OutputDescription {
+            cv,
+            cmu,
+            ephemeral_key,
+            enc_ciphertext,
+            out_ciphertext,
+            zkproof,
+        })
+    }
+
+    pub fn write<W: Write>(&self, mut writer: W) -> io::Result<()> {
+        self.cv.write(&mut writer)?;
+        self.cmu.into_repr().write_le(&mut writer)?;
+        self.ephemeral_key.write(&mut writer)?;
+        writer.write_all(&self.enc_ciphertext)?;
+        writer.write_all(&self.out_ciphertext)?;
+        writer.write_all(&self.zkproof)
+    }
+}
+
+enum SproutProof {
+    Groth([u8; GROTH_PROOF_SIZE]),
+    PHGR([u8; PHGR_PROOF_SIZE]),
+}
+
+pub struct JSDescription {
+    vpub_old: Amount,
+    vpub_new: Amount,
+    anchor: [u8; 32],
+    nullifiers: [[u8; 32]; ZC_NUM_JS_INPUTS],
+    commitments: [[u8; 32]; ZC_NUM_JS_OUTPUTS],
+    ephemeral_key: [u8; 32],
+    random_seed: [u8; 32],
+    macs: [[u8; 32]; ZC_NUM_JS_INPUTS],
+    proof: SproutProof,
+    ciphertexts: [[u8; 601]; ZC_NUM_JS_OUTPUTS],
+}
+
+impl JSDescription {
+    pub fn read<R: Read>(mut reader: R, use_groth: bool) -> io::Result<Self> {
+        let vpub_old = Amount(reader.read_i64::<LittleEndian>()?);
+        let vpub_new = Amount(reader.read_i64::<LittleEndian>()?);
```

Comment:
Technically these fields are unsigned (section 7.2 of the spec). More precisely they're in the range {0..MAX_MONEY}.

Reply:
`Amount` parallels `CAmount` in `zcashd`, handling the fact that amounts may be subtracted with negative results. I'll add `impl From<u64> for Amount` which performs that check during conversion, to make things clearer (so we use `read_u64()` to more closely match the spec).

Reply:
Maybe there should be a `RestrictedAmount` or `CheckedAmount` trait that is guaranteed to be in the {0..MAX_MONEY} range, with checked arithmetic?